### PR TITLE
[jvm-packages]Adds the ability to pass multiple custom evaluation

### PR DESCRIPTION
**Pull request by Helw150**
Putting this here to gain access to the CI. This PR is based off updating and old fork of XGBoost which added this functionality to the current master state, so may have errors from merge conflict.

### CODE REVIEW

**Comment by Helw150** [comment]
Additionally, I ran ScalaFmt and google-java-fmt - which passed the XGBoost official style checks, but I am not sure if those meet the standards of XGBoost open source in practice.

**Comment by hcho3** [comment]
This is great. Does your fork also have watchlist feature for XGBoost4J-Spark?

**Comment by CodingCat** [comment]
would you please avoid changing format in the same PR, though there are some style problems especially for Java code, 

we may focus on your watch list thing in this PR

**Comment by Helw150** [comment]
@CodingCat I just styled with `google-java-format` in order to pass the linter as quickly as possible rather than doing manual nits. I can try to revert as many style changes as possible if you'd like.

**Comment by Helw150** [comment]
Alternatively, let me know if you guys have a preferred formatter or a more specific ruleset of your linter I should follow. I try to leave as much as possible to linters and formatters style wise.

**Comment by CodingCat** [comment]
I don't think the current linter is checking jvm/* java has no style checker now...scala is using scala-style-check plugin....

**Comment by Helw150** [comment]
Do you guys have a style guide then? I'm working on this as a side project, so haven't particularly set up a good dev environment for iterating on this so anything that simplifies my process to follow y'alls ruleset is much appreciated!

**Comment by Helw150** [comment]
The java did seem to be linting with Checkstyle(that was my motivation for using a formatter), so I just adhered to the linting rules Checkstyle imposed.

**Comment by CodingCat** [comment]
what's the error you received? if you don't change any format, I believe it will not break with stylistic errors (as all the other PRs are successfully passing tests)

**Comment by CodingCat** [comment]
our scala style is essentially a subset (removed some spark related thing) of https://github.com/databricks/scala-style-guide

**Comment by Helw150** [comment]
@CodingCat 
```
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[108] (sizes) LineLength: Line is longer than 100 characters (found 114).
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[139] (indentation) Indentation: 'method def' child have incorrect indentation level 6, expected level should be 4.
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[139] (sizes) LineLength: Line is longer than 100 characters (found 119).
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[207] (indentation) Indentation: 'for' child have incorrect indentation level 18, expected level should be 12.
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[207] (indentation) Indentation: 'member def type' have incorrect indentation level 18, expected level should be 12.
```

I'll just fix these manually. Running the Google Style guide formatter fixes all these Lint issues, but your style guide in practice seems to differ from the linter.

**Comment by CodingCat** [comment]
still confused why all the other PRs didn't touch java can pass lint checker (are you referring to Travis CI's lint checker), but you have such an issue

can you just put stylistic issue aside and implement functionalities? or just follow what it says about style?

**Comment by CodingCat** [comment]
maybe lint works as "check when there is any modification <b> in the directory</b>" since you add a new file there

**Comment by Helw150** [comment]
@CodingCat @hcho3 I don't believe the PR-merge CI errors stem from my changes as it stems from a C error and seems to be affecting a few other open PR's. Am I correct in that?

**Comment by hcho3** [comment]
@Helw150 Correct, I'm currently looking into it.

**Comment by Helw150** [comment]
@CodingCat You seem to be the one to look to for review on this one? Whenever you get a chance to review, I'd be super grateful :)

**Comment by Helw150** [comment]
Ping @CodingCat or @hcho3 for review :)

**Comment by CodingCat** [comment]
@Helw150 we are focusing on 0.8 release, will look at it later

**Comment by Helw150** [comment]
Great, just making sure it made it on your radar. I'll add change `WIP`->`Ready For Reminder`. Thanks for the help

**Comment by Helw150** [comment]
@CodingCat @hcho3 Just a note, Next week is the last week of the internship where I have been working on adding this functionality as a side project. I would really appreciate it if one of you has a bit of time to review.

Sorry to ping again 😓

**Comment by CodingCat** [comment]
Sorry, No guarantee on that, but we will try....one thing to note is that none of us work on this as full time, instead, we spent time which we squeeze from our daily jobs

**Comment by Helw150** [comment]
Yup, totally understand as it's the same for me on this PR. Just wanted to make sure it was still on your radar as I had seen new PR's get reviewed and merged as I know it's easy to forget things when they aren't on your roadmap!

If it can't get reviewed, no worries. It's living happily on our fork internally, so for our purposes no harm no foul if we can't get this merged.

**Comment by hcho3** [comment]
@Helw150 We do aim to get all pull requests merged eventually, so that the list doesn't exceed on page.

@CodingCat Wish I knew a bit about Scala so that I can help you here. The Java land is still quite new to me hehe

**Comment by Helw150** [comment]
@hcho3 No worries. I'm in no huge rush as long as it's on your list 👍

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java [hunk 1]
```diff
@@ -187,11 +208,16 @@ public static Booster train(
       //evaluation
       if (evalMats.length > 0) {
         float[] metricsOut = new float[evalMats.length];
-        String evalInfo;
-        if (eval != null) {
-          evalInfo = booster.evalSet(evalMats, evalNames, eval, metricsOut);
+        String evalInfo = "";
+        if (evals != null && evals.length > 0) {
+          for (int i = 0; i < evals.length; i++) {
+            String evalLine = booster.evalSet(evalMats, evalNames, evals[i], metricsOut);
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java]: isn't it replacing metricsOut for every time? 

and metricsOut is used in early stopping, etc. which makes it problematic

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 2]
```diff
@@ -18,14 +18,57 @@ package ml.dmlc.xgboost4j.scala
...
 import java.io.InputStream
 
-import ml.dmlc.xgboost4j.java.{Booster => JBooster, XGBoost => JXGBoost, XGBoostError}
+import ml.dmlc.xgboost4j.java.{
+  Booster => JBooster,
+  XGBoost => JXGBoost,
+  XGBoostError,
+  BoosterResults,
+  IEvaluation
+}
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: unnecessary multi-lining

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 3]
```diff
@@ -18,14 +18,57 @@ package ml.dmlc.xgboost4j.scala
...
+      dtrain: DMatrix,
+      params: Map[String, Any],
+      round: Int,
+      watches: Map[String, DMatrix] = Map[String, DMatrix](),
+      metrics: Array[Array[Float]] = null,
+      obj: ObjectiveTrait = null,
+      evals: Array[IEvaluation] = null,
+      earlyStoppingRound: Int = 0,
+      booster: Booster = null
+  ): BoosterResults = {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: consolidate line 45 and 46

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 4]
```diff
@@ -18,14 +18,57 @@ package ml.dmlc.xgboost4j.scala
...
+    } else {
+      booster.booster
+    }
+    val xgboostResults = JXGBoost.trainWithResults(
+      dtrain.jDMatrix,
+      // we have to filter null value for customized obj and eval
+      params
+        .filter(_._2 != null)
+        .mapValues(_.toString.asInstanceOf[AnyRef])
+        .asJava,
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: it's better to have a new variable for filtered params, to make code cleaner

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 5]
```diff
@@ -18,14 +18,57 @@ package ml.dmlc.xgboost4j.scala
...
+      round,
+      jWatches,
+      metrics,
+      obj,
+      evals,
+      earlyStoppingRound,
+      jBooster
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: I don't think you need make one parameter per line

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java [hunk 6]
```diff
@@ -136,12 +137,32 @@ public static Booster train(
           IEvaluation eval,
           int earlyStoppingRound,
           Booster booster) throws XGBoostError {
+    if(eval != null) {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java]: space after `if`

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java [hunk 7]
```diff
@@ -136,12 +137,32 @@ public static Booster train(
...
+  public static BoosterResults trainWithResults(
+      DMatrix dtrain,
+      Map<String, Object> params,
+      int round,
+      Map<String, DMatrix> watches,
+      float[][] metrics,
+      IObjective obj,
+      IEvaluation[] evals,
+      int earlyStoppingRound,
+      Booster booster) throws XGBoostError {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java]: it looks like you have empty evals and null both indicate no customized evaluation function

to make the logic clear, we'd make only one of them to be a valid input and the other triggers an exception

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 8]
```diff
@@ -52,22 +95,80 @@ object XGBoost {
...
+    val xgboostResults = trainWithResults(
+      dtrain,
+      params,
+      round,
+      watches,
+      metrics,
+      obj,
+      evals,
+      earlyStoppingRound,
+      booster
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: unnecessary multi-lining

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 9]
```diff
@@ -52,22 +95,80 @@ object XGBoost {
       watches: Map[String, DMatrix] = Map(),
       metrics: Array[Array[Float]] = null,
       obj: ObjectiveTrait = null,
-      eval: EvalTrait = null,
+      eval: IEvaluation = null,
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: EvalTrait is not used at all?

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 10]
```diff
@@ -94,11 +195,20 @@ object XGBoost {
       nfold: Int = 5,
       metrics: Array[String] = null,
       obj: ObjectiveTrait = null,
-      eval: EvalTrait = null): Array[String] = {
+      eval: EvalTrait = null
+  ): Array[String] = {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: consolidate 198 and 199

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 11]
```diff
@@ -52,22 +95,80 @@ object XGBoost {
...
+    val xgboostResults = trainWithResults(
+      dtrain,
+      params,
+      round,
+      watches,
+      metrics,
+      obj,
+      evals,
+      earlyStoppingRound,
+      booster
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: unnecessary multi-lining

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala [hunk 12]
```diff
@@ -52,22 +95,80 @@ object XGBoost {
...
+      dtrain: DMatrix,
+      params: Map[String, Any],
+      round: Int,
+      watches: Map[String, DMatrix] = Map[String, DMatrix](),
+      metrics: Array[Array[Float]] = null,
+      obj: ObjectiveTrait = null,
+      evals: Array[IEvaluation] = null,
+      earlyStoppingRound: Int = 0,
+      booster: Booster = null
+  ): Booster = {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala]: consolidate 157 and 158

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java (hunk 1 above)
Comment (Helw150) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java]: Hmm, yeah. For early stopping, I'd probably need to add either an assumption about which eval metric is used, or a way for the user to choose which one matters. 

Are you fine with the first metric in the array of evaluations being used as the early stopping metric?

**Comment by CodingCat** [comment]
any update on this?

**Comment by Helw150** [comment]
@CodingCat As I said a bit ago, the internship I was working on this as a side project for ended so I don't have nearly as much time to make fixes for the next couple of months. I'll try to make incremental changes based on your comments over the next few weeks, but I don't have any explicit time on my calendar to work on this while at school.

**Comment by hcho3** [comment]
Please ignore the failed test `continuous-integration/jenkins/pr-head`. (This was due to misconfiguration of Jenkins CI server.) Only `continuous-integration/jenkins/pr-merge` is relevant.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java (hunk 1 above)
Comment (Helw150) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java]: Ok, disabled early stopping for multiple eval functions. I chose to simply print via rabit and continue training, but let me know if you'd prefer something else (namely a full error to be thrown).

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java (hunk 7 above)
Comment (Helw150) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java]: If there are no evaluation functions, I chose to allow null and throw error on empty array of evaluation fucntions.

**Comment by Helw150** [comment]
@CodingCat This should be in a state for you to give another look. I fixed all the issues you had, the primary thing I am uncertain on is how I should handle the one of the lines you said shouldn't have one argument per line. 

The [style guide](https://github.com/databricks/scala-style-guide#spacing-and-indentation) you linked earlier has the following as it's standard for long method invocations, so for now I've reverted line 54-64 to a single parameter per line to meet the style guide. 

- For method and class constructor invocations, use 2 space indentation for its parameters and put each in each line when the parameters don't fit in two lines.

  ```scala
  foo(
    someVeryLongFieldName,  // 2 space indent here
    andAnotherVeryLongFieldName,
    "this is a string",
    3.1415)

  new Bar(
    someVeryLongFieldName,  // 2 space indent here
    andAnotherVeryLongFieldName,
    "this is a string",
    3.
```

**Comment by hcho3** [comment]
Closing as stale

**Pull request closed without merging**